File: frontend-src/src-tauri/src/services/preferences.rs
use rusqlite::{params, Connection};
use serde_json::{json, Value};

use crate::db;
use crate::error::ApiError;

use super::merge_object;

// Singleton (id = 1) -- mismo criterio que app_preferences en el backend
// web. GET siempre responde algo (valores por defecto si la fila no existe
// todavía), PUT hace upsert.
//
// teacher_name/teacher_profile/teacher_notes: Perfil Docente (mismos campos
// que api/app/migrations/0014_teacher_profile, 0019_teacher_notes,
// 0020_teacher_personal_data en el backend web) -- se habían dejado fuera
// del primer paso a escritorio, ver migrations/0008_teacher_profile_and_sa_fields.sql.
// teacherHasPhoto es un campo CALCULADO (teacher_photo IS NOT NULL), igual
// que en el backend web -- la foto en sí viaja aparte (ver get/set/delete_photo
// más abajo), nunca dentro de este JSON.
pub fn get(conn: &Connection) -> Result<Value, ApiError> {
    let row = conn.query_row(
        "SELECT layout_mode, default_calendar_view, grade_scale, teacher_name, teacher_profile, teacher_notes, teacher_photo IS NOT NULL FROM app_preferences WHERE id = 1",
        [],
        |row| {
            let layout_mode: Option<String> = row.get(0)?;
            let default_calendar_view: Option<String> = row.get(1)?;
            let grade_scale: String = row.get(2)?;
            let teacher_name: String = row.get(3)?;
            let teacher_profile: String = row.get(4)?;
            let teacher_notes: String = row.get(5)?;
            let teacher_has_photo: bool = row.get(6)?;
            Ok((layout_mode, default_calendar_view, grade_scale, teacher_name, teacher_profile, teacher_notes, teacher_has_photo))
        },
    );

    match row {
        Ok((layout_mode, default_calendar_view, grade_scale, teacher_name, teacher_profile, teacher_notes, teacher_has_photo)) => Ok(json!({
            "layoutMode": layout_mode,
            "defaultCalendarView": default_calendar_view,
            "gradeScale": serde_json::from_str::<Value>(&grade_scale).unwrap_or_else(|_| json!([])),
            "teacherName": teacher_name,
            "teacherProfile": serde_json::from_str::<Value>(&teacher_profile).unwrap_or_else(|_| json!([])),
            "teacherNotes": teacher_notes,
            "teacherHasPhoto": teacher_has_photo,
        })),
        Err(rusqlite::Error::QueryReturnedNoRows) => Ok(json!({
            "layoutMode": Value::Null,
            "defaultCalendarView": Value::Null,
            "gradeScale": Value::Array(vec![]),
            "teacherName": "",
            "teacherProfile": Value::Array(vec![]),
            "teacherNotes": "",
            "teacherHasPhoto": false,
        })),
        Err(e) => Err(e.into()),
    }
}

pub fn update(conn: &Connection, body: Value) -> Result<Value, ApiError> {
    let current = get(conn)?;
    let merged = merge_object(&current, &body);

    let layout_mode = merged.get("layoutMode").and_then(Value::as_str);
    let default_calendar_view = merged.get("defaultCalendarView").and_then(Value::as_str);
    let grade_scale = merged.get("gradeScale").cloned().unwrap_or_else(|| json!([]));
    let grade_scale_str = serde_json::to_string(&grade_scale).map_err(ApiError::internal)?;
    let teacher_name = merged.get("teacherName").and_then(Value::as_str).unwrap_or("");
    let teacher_profile = merged.get("teacherProfile").cloned().unwrap_or_else(|| json!([]));
    let teacher_profile_str = serde_json::to_string(&teacher_profile).map_err(ApiError::internal)?;
    let teacher_notes = merged.get("teacherNotes").and_then(Value::as_str).unwrap_or("");

    // teacher_photo/teacher_photo_content_type quedan fuera a propósito de
    // este INSERT/UPDATE (ON CONFLICT no los toca): viajan por
    // get/set/delete_teacher_photo, nunca por aquí -- mismo motivo que
    // teacherHasPhoto es de solo lectura en el JSON de arriba.
    conn.execute(
        "INSERT INTO app_preferences (id, layout_mode, default_calendar_view, grade_scale, teacher_name, teacher_profile, teacher_notes, updated_at)
         VALUES (1, ?, ?, ?, ?, ?, ?, ?)
         ON CONFLICT (id) DO UPDATE SET
            layout_mode = excluded.layout_mode,
            default_calendar_view = excluded.default_calendar_view,
            grade_scale = excluded.grade_scale,
            teacher_name = excluded.teacher_name,
            teacher_profile = excluded.teacher_profile,
            teacher_notes = excluded.teacher_notes,
            updated_at = excluded.updated_at",
        params![layout_mode, default_calendar_view, grade_scale_str, teacher_name, teacher_profile_str, teacher_notes, db::now_iso()],
    )?;
    get(conn)
}

// Foto de perfil del profesor: mismo patrón que services/photos.rs (fotos
// de alumnado) pero sobre la fila única de preferencias en vez de por id.
// Si la fila de app_preferences todavía no existe (nunca se ha guardado
// nada), la crea con sus DEFAULT antes de fijar la foto -- a diferencia de
// update() de arriba, aquí no hay un "current" que journalear primero.
pub fn get_photo(conn: &Connection) -> Result<Option<(Vec<u8>, String)>, ApiError> {
    let result = conn.query_row(
        "SELECT teacher_photo, teacher_photo_content_type FROM app_preferences WHERE id = 1 AND teacher_photo IS NOT NULL",
        [],
        |row| {
            let bytes: Vec<u8> = row.get(0)?;
            let content_type: Option<String> = row.get(1)?;
            Ok((bytes, content_type))
        },
    );
    match result {
        Ok((bytes, content_type)) => Ok(Some((
            bytes,
            content_type.unwrap_or_else(|| "application/octet-stream".to_string()),
        ))),
        Err(rusqlite::Error::QueryReturnedNoRows) => Ok(None),
        Err(e) => Err(e.into()),
    }
}

pub fn set_photo(conn: &Connection, bytes: Vec<u8>, content_type: &str) -> Result<(), ApiError> {
    conn.execute(
        "INSERT INTO app_preferences (id, grade_scale, updated_at, teacher_photo, teacher_photo_content_type)
         VALUES (1, '[]', ?, ?, ?)
         ON CONFLICT (id) DO UPDATE SET
            teacher_photo = excluded.teacher_photo,
            teacher_photo_content_type = excluded.teacher_photo_content_type,
            updated_at = excluded.updated_at",
        params![db::now_iso(), bytes, content_type],
    )?;
    Ok(())
}

pub fn delete_photo(conn: &Connection) -> Result<(), ApiError> {
    conn.execute(
        "UPDATE app_preferences SET teacher_photo = NULL, teacher_photo_content_type = NULL, updated_at = ? WHERE id = 1",
        params![db::now_iso()],
    )?;
    Ok(())
}

File: api/app/services/preferences.py
from typing import Optional

from psycopg.types.json import Json

from services.db import get_conn
from services.schemas import ApiModel


class PreferencesInput(ApiModel):
    layout_mode: Optional[str] = None
    default_calendar_view: Optional[str] = None
    grade_scale: Optional[list] = None
    # Rasgos de estilo docente (p.ej. "Cercano y motivador", "Prioriza la
    # práctica sobre la teoría"...) -- se inyectan en el prompt de cada SA
    # generada con IA para que escriba coherente con cómo enseña este
    # profesor, no con un "eres un profesor" genérico.
    teacher_profile: Optional[list] = None
    # Notas libres complementarias a teacher_profile -- preferencias sobre
    # el material en sí (formato, extensión, tono...) que no encajan como
    # una etiqueta corta. Se inyectan también en el prompt.
    teacher_notes: Optional[str] = None
    # Datos personales del profesor (nombre para mostrar en la app; la foto
    # va aparte, mismo patrón BYTEA que students.foto -- ver services/
    # photos.py -- con su propio endpoint binario en vez de viajar aquí).
    teacher_name: Optional[str] = None


class Preferences(PreferencesInput):
    grade_scale: list = []
    teacher_profile: list = []
    teacher_notes: str = ''
    teacher_name: str = ''
    # Solo lectura (no en PreferencesInput): evita que el frontend tenga que
    # intentar cargar /preferences/photo a ciegas y detectar el 404 -- sabe
    # de antemano si hay algo que pedir.
    teacher_has_photo: bool = False


# Singleton (id = true, ver DDL) — igual que app_db en el sistema viejo. Se
# trata como "siempre hay una fila lógica" aunque todavía no se haya escrito
# ninguna: GET devuelve valores por defecto, PUT crea la fila si hace falta.
def get_preferences() -> Preferences:

    with get_conn() as conn:

        with conn.cursor() as cur:

            cur.execute(
                "SELECT layout_mode, default_calendar_view, grade_scale, teacher_profile, teacher_notes, teacher_name, "
                "(teacher_photo IS NOT NULL) AS teacher_has_photo FROM app_preferences WHERE id = true"
            )

            row = cur.fetchone()

            return Preferences.model_validate(row) if row else Preferences()


def update_preferences(data: PreferencesInput) -> Preferences:

    current = get_preferences()

    merged = current.model_copy(update=data.model_dump(exclude_unset=True))

    with get_conn() as conn:

        with conn.cursor() as cur:

            cur.execute(
                """
                INSERT INTO app_preferences (id, layout_mode, default_calendar_view, grade_scale, teacher_profile, teacher_notes, teacher_name, updated_at)
                VALUES (true, %s, %s, %s, %s, %s, %s, now())
                ON CONFLICT (id) DO UPDATE SET
                    layout_mode = EXCLUDED.layout_mode,
                    default_calendar_view = EXCLUDED.default_calendar_view,
                    grade_scale = EXCLUDED.grade_scale,
                    teacher_profile = EXCLUDED.teacher_profile,
                    teacher_notes = EXCLUDED.teacher_notes,
                    teacher_name = EXCLUDED.teacher_name,
                    updated_at = EXCLUDED.updated_at
                """,
                [merged.layout_mode, merged.default_calendar_view, Json(merged.grade_scale), Json(merged.teacher_profile), merged.teacher_notes, merged.teacher_name]
            )

    return merged


# Foto del profesor, mismo patrón BYTEA que students.foto (services/
# photos.py) pero sobre la fila singleton de app_preferences en vez de por
# id -- upsert en set/delete porque, a diferencia de students, la fila de
# preferencias puede no existir todavía la primera vez que se sube una foto.
def get_teacher_photo() -> Optional[tuple[bytes, str]]:

    with get_conn() as conn:

        with conn.cursor() as cur:

            cur.execute("SELECT teacher_photo, teacher_photo_content_type FROM app_preferences WHERE id = true")

            row = cur.fetchone()

            if row is None or row["teacher_photo"] is None:
                return None

            return bytes(row["teacher_photo"]), row["teacher_photo_content_type"]


def set_teacher_photo(data: bytes, content_type: str) -> None:

    with get_conn() as conn:

        with conn.cursor() as cur:

            cur.execute(
                """
                INSERT INTO app_preferences (id, teacher_photo, teacher_photo_content_type, updated_at)
                VALUES (true, %s, %s, now())
                ON CONFLICT (id) DO UPDATE SET
                    teacher_photo = EXCLUDED.teacher_photo,
                    teacher_photo_content_type = EXCLUDED.teacher_photo_content_type,
                    updated_at = EXCLUDED.updated_at
                """,
                [data, content_type]
            )


def delete_teacher_photo() -> None:

    with get_conn() as conn:

        with conn.cursor() as cur:

            cur.execute(
                """
                INSERT INTO app_preferences (id, teacher_photo, teacher_photo_content_type, updated_at)
                VALUES (true, NULL, NULL, now())
                ON CONFLICT (id) DO UPDATE SET
                    teacher_photo = NULL,
                    teacher_photo_content_type = NULL,
                    updated_at = now()
                """
            )
